### PR TITLE
[DF] Change the signature of RInterface::Describe (Fix #8893)

### LINKED ISSUE #8893: [DF] `Describe` output does not display correctly in Jupyter notebooks
image

### PR DESCRIPTION
**Pull request by ikabadzhov**
# This Pull request:

[DF] Change the signature of RInterface::Describe

## Changes or fixes:

A new structure `DFDescription` is introduced.
It has 2 member strings, corresponding to the brief and the full description.
It allows more interactive output of these strings.

`RInterface::Describe` now returns a `DFDescription`.
As brief description is the output from `RInterface::DescribeDataset`.
As full description is the remaining code from `RInterface::Describe`.
Moreover, `RInterface::DescribeDataset` is now a private method.

RDFDescription has the following methods:
* `AsString(bool)` -> returning brief/full description as a string
* `Print(bool)` -> printing the content of `AsString(bool)`
* overloaded `<<` -> returns ostream corresponding to `AsString(shortFormat=false)`
* printValue -> returns string corresponding to AsString(shortFomat=false)
* `__repr__` pythonization -> assigning `__repr__` to `AsString(shortFormat=false)`

Tests and Tutorials were adapted correspondingly.

## Checklist:

- [ ] tested changes locally
- [ ] updated the docs (if necessary)

This PR fixes #8893

### CODE REVIEW

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 1]
```diff
@@ -118,6 +119,91 @@ class RInterface {
    /// Contains the columns defined up to this node.
    RDFInternal::RColumnRegister fColRegister;
 
+   /// \brief Get descriptive information about the dataset.
+   /// \return Info describing the dataset as a multi-line string
+   ///
+   /// The information returned by this convenience function is meant for interactive
+   /// use. The exact string format should not be parsed automatically and can be subject to change.
+   ///
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:127]: doesn't apply anymore

```suggestion
```

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 2]
```diff
@@ -2150,7 +2236,7 @@ public:
    /// for (auto &&colName : colNames) std::cout << colName << std::endl;
    /// ~~~
    ///
-   ColumnNames_t GetColumnNames()
+   ColumnNames_t GetColumnNames() const
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2239]: this is unrelated to the PR, and should be discussed carefully: it's a change in RDataFrame's public interface, and a `const` method promises to be thread-safe, which should be checked. Please remove for now.

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 3]
```diff
@@ -2192,7 +2278,7 @@ public:
    /// std::cout << "Column " << colType << " has type " << colType << std::endl;
    /// ~~~
    ///
-   std::string GetColumnType(std::string_view column)
+   std::string GetColumnType(std::string_view column) const
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2281]: same as above

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 4]
```diff
@@ -2227,15 +2313,14 @@ public:
    /// RDataFrame df(10);
    /// auto df2 = df.Define("x", "1.f").Define("s", "\"myStr\"");
    /// // Describe the dataframe
-   /// std::cout << df2.Describe() << std::endl;
+   /// df2.Describe().Print()
+   /// df2.Describe().Print(/*shortFormat =*/true)
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2317]:
```suggestion
   /// df2.Describe().Print(/*shortFormat=*/true)
```

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 5]
```diff
@@ -2227,15 +2313,14 @@ public:
    /// RDataFrame df(10);
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2313]: must update the method description above, e.g. it shouldn't point users to `DescribeDataset`

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 6]
```diff
@@ -2322,7 +2409,7 @@ public:
    /// for (auto &&defColName : defColNames) std::cout << defColName << std::endl;
    /// ~~~
    ///
-   ColumnNames_t GetDefinedColumnNames()
+   ColumnNames_t GetDefinedColumnNames() const
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2412]: same as above

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 7]
```diff
@@ -2764,7 +2760,7 @@ private:
       fColRegister.AddAlias("tdfslot_", slotColName);
    }
 
-   std::vector<std::string> GetColumnTypeNamesList(const ColumnNames_t &columnList)
+   std::vector<std::string> GetColumnTypeNamesList(const ColumnNames_t &columnList) const
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2763]: same as above

File: tutorials/dataframe/df033_Describe.py [hunk 8]
```diff
@@ -17,7 +17,9 @@
 
 # Describe the state of the dataframe.
 # Note that this operation is not running the event loop.
-print(df.Describe())
+# It could be done in 2 ways - either by returning a string, or directly printing.
+print(df.Describe().AsString())
+df.Describe().Print()
```
Comment (eguiraud) [inline tutorials/dataframe/df033_Describe.py:22]: more future-proof:

```suggestion
# Describe returns a DFDescription object, which has e.g. a Print method. See its docs for more information.
df.Describe().Print()
```

File: tutorials/dataframe/df033_Describe.py [hunk 9]
```diff
@@ -32,4 +34,6 @@
     df.Mean('Dimuon_mass').GetValue()))
 
 # Describe again the state of the dataframe.
-print(df.Describe())
+# This time True argument is added, which corresponds to printing the brief description only.
+print(df.Describe().AsString(True))
+df.Describe().Print(True)
```
Comment (eguiraud) [inline tutorials/dataframe/df033_Describe.py:39]:
```suggestion
# This time we ask for the `shortFormat`, which only prints a brief description of the dataset:
df.Describe().Print(shortFormat=True)
```

**Review by eguiraud — CHANGES_REQUESTED** [review]
The design is correct, some notes on the implementation below.

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx (hunk 2 above)
Comment (ikabadzhov) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2239]: I will remove the `const` (here and on the methods below), but note that in this case `Describe()` can no longer be `const`.

**Comment by vepadulano** [comment]
Thanks @ikabadzhov for this very nice improvements! +1 for the usage of `const` whenever needed. Do we also want to include the overload of the `operator<<` and the pythonization for the `__repr__` and `__str__` dunder methods in this release or we leave them for a next time?

**Comment by ikabadzhov** [comment]
Thank you @vepadulano. As a current solution, we decided that `Describe()` returns a `RDFDescribe` object which has 2 methods `Print(bool)` and `AsString(bool)`, in stead of overloading `operator<<` and its `std::string()`.
I see now that alternatively, we could overload these operators of the `RDFDescribe` class.

Reason to think that current solution with separate methods is better is the design that `Print` and `AsString` methods take one input argument (`bool` to determine if we want short or full description).

Given your comment, I see it as a valid argument to have: `RDFDescribe Describe(bool)`, and then `RDFDescribe` has overloaded `<<` and `std::string()`. What do you think about that?

**Comment by eguiraud** [comment]
Good point, yes I think we want `__repr__` for a nicer experience in notebooks and REPLs (https://github.com/root-project/root/issues/8893), as well as `std::string printValue(RDFDescription *tdf)` for the cling REPL -- for the `RDFDescription` class, indeed.

@ikabadzhov can you take care of it? For the `printValue` function you can see how it's done e.g. in `RDataFrame.hxx`, for the `__repr__` pythonization I think Vincenzo can give you some pointers.

**Comment by ikabadzhov** [comment]
@eguiraud @vepadulano if we have overloaded `<<`, `string`, `__repr__`, do we then need to keep `Print(bool)` and `AsString(bool)`? I think no, or?

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx (hunk 5 above)
Comment (ikabadzhov) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2306]: fixed in the latest commit

File: bindings/pyroot/pythonizations/test/CMakeLists.txt [hunk 10]
```diff
@@ -108,6 +108,11 @@ if (dataframe)
     endif()
 endif()
 
+# RDFDescription pythonization
+if (rdfdescription)
```
Comment (etejedor) [inline bindings/pyroot/pythonizations/test/CMakeLists.txt:112]: Shouldn't this be under the `if (dataframe)` condition block above?

Comment (vepadulano) [reply bindings/pyroot/pythonizations/test/CMakeLists.txt:112]: yes indeed there is no `rdfdescription` cmake option so the test should go with the other rdf tests

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdfdescription.py [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+# All rights reserved.                                                         #
+#                                                                              #
+# For the licensing terms see $ROOTSYS/LICENSE.                                #
+# For the list of contributors see $ROOTSYS/README/CREDITS.                    #
+################################################################################
+
+from . import pythonization
+
+@pythonization("RDFDescription", ns="ROOT::RDF")
+def pythonize_rdataframe(klass):
```
Comment (etejedor) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdfdescription.py:14]: Congrats for your first pythonization!

I would call this `pythonize_rdfdescription` to be consistent in the naming.

Comment (ikabadzhov) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdfdescription.py]: corrected in the latest commit (and also the other typo)

File: bindings/pyroot/pythonizations/test/rdfdescription.py [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+class RDFDescriptionTest(unittest.TestCase):
+    """
+    Testing of RDFDescription pythonization
+    """
+    def test_repr(self):
+        """
+        Test supported __repr__
+        """
+        df1 = ROOT.RDataFrame(1);
+        ref1 = ("Empty dataframe filling 1 row\n"
```
Comment (etejedor) [inline bindings/pyroot/pythonizations/test/rdfdescription.py:13]: Assuming that there are tests in RDataFrame for `AsString` that check it does the right thing, this test could be simply `         self.assertEqual(df1.Describe().AsString(), repr(df1.Describe()));`, i.e. no need to test again what is tested in the `AsString` tests.

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 13]
```diff
@@ -2220,22 +2299,21 @@ public:
...
    /// // Describe the dataframe
-   /// std::cout << df2.Describe() << std::endl;
+   /// df2.Describe().Print()
+   /// df2.Describe().Print(/*shortFormat=*/true)
+   /// std::cout << df2.Describe().AsString() << std::endl;
+   /// std::cout << df2.Describe().AsString(/*shortFormat=*/true) << std::endl;
    /// ~~~
    ///
-   std::string Describe()
+   RDFDescription Describe()
```
Comment (etejedor) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2315]: Why `RDFDescription` and not `RDescription` as in `RDisplay`? The namespace `ROOT::RDF` already qualifies it as an RDF thing no?

Comment (etejedor) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2315]: Yes, I'm asking if there's a reason to name it `RDFDescription` and not `RDescription`.

Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2315]: Yes, we might have other descriptions of things still under the `ROOT::RDF` namespace, and in Doxygen (or when searching the code in general) it's better to search for the `RDFDescription` class than for a generic `RDescription` class across all ROOT.

File: bindings/pyroot/pythonizations/test/rdfdescription.py (hunk 12 above)
Comment (ikabadzhov) [reply bindings/pyroot/pythonizations/test/rdfdescription.py:13]: I am applying this change and creating a new commit now. Everything else I keep unchanged (so far).

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx (hunk 13 above)
Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2315]: mmh on second thought I don't know, Enric is right that we have some seriously generic names in the `ROOT::RDF` namespace, like `RDisplay` and `RInterface`.

I guess the question is just whether we'll ever have another description of something else, like a `RDF::RFilterDescription` :thinking:

Comment (etejedor) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2315]: > I guess the question is just whether we'll ever have another description of something else, like a RDF::RFilterDescription

In that case, for our peace of mind, we can argue that `ROOT::RDF::RDescription` means the description of the RDF (the `RDF` namespace qualifies it as such), if the description is of an RDF filter, it is fair that it's called `ROOT::RDF:RFilterDescription`

Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2315]: Peace of mind would be an argument for leaving the more specific name :smile:

Ivan, Enric, pick the one you like best, I have no preference.

Comment (ikabadzhov) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2315]: I think we should also think if later on, we will have other `RDF` "instances" which are used to "present" some specific part of RDF itself. As an example potentially `RDFMetaData`.
I believe answer should converge to yes, and hence I would go for `RDFDescription`. Moreover, I just saw that there is a `RDFTypeNameGetter` class, so keeping `RDF` seem consistent.

Remark, no matter which version we choose, very likely the other version is never going to be used (mainly for consistency and avoiding confusion - might be obvious but I want to make sure it is considered).

**Review by eguiraud — APPROVED** [review]
LGTM

**Review by vepadulano — APPROVED** [review]
Very nice!
